### ISSUE TITLE
Double archiving

### ISSUE THREAD

**Issue by GeneralNotability** [GeneralNotability/spihelper#14]
Reported by JJMC89: https://en.wikipedia.org/wiki/Wikipedia:Sockpuppet_investigations/14.0.180.170/Archive?action=history archived the same section twice for some reason.

**Comment by roysmith** [comment]
Two other examples:

https://en.wikipedia.org/w/index.php?title=Wikipedia:Sockpuppet_investigations/JohnGotten/Archive&type=revision&diff=1010497534&oldid=1008672694&diffmode=source

And the first two revisions in https://en.wikipedia.org/w/index.php?title=Wikipedia:Sockpuppet_investigations/BoommBoomm87/Archive&action=history

In all three cases, it's two separate revisions, a few seconds apart.

**Comment by GeneralNotability** [comment]
Yeah, this is a stumper. This sure looks like a race condition involving multiple section-archive operations running at once, but I can't see anywhere that a race condition could be happening - all of the API operations are using `await`. I'm pretty sure everyone hitting this bug is using the one-click archive function.

**Comment by roysmith** [comment]
Why do you think one-click archiver?  The commit messages ("Archiving case section from...") look like they're coming from spiHelper_archiveCaseSection().

**Comment by roysmith** [comment]
I have an idea what's going on.  In  spiHelper_archiveCase(), you basically do:

```
for section in caseSections {
    spiHelper_archiveCaseSection()
    spiHelper_getInvestigationSectionIDs()
}
```

where `spiHelper_archiveCaseSection()` calls `spiHelper_editPage()` and `spiHelper_getInvestigationSectionIDs()` calls `api.get()` and reparses the page.  Even with the awaits on the various calls, I don't think the MW API guarantees that the api.get() will see the version of the page you just updated with the editPage().  You could be hitting two different server processes with inconsistent views of the page due to caching and/or database replication.

My suggestion is to have the api.get(action: 'parse'...) call include revid in prop.  My hunch is you'll find once in a while, you'll re-fetch the same revid that you had before.  I think that would result in the behavior you're seeing.  For now, maybe just log the revid in the edit comment and the next time this happens, we'll see if they're the same for the duplicate edits on the archive.

**Comment by GeneralNotability** [comment]
Re "one-click archiver," sorry, misspoke - meant the "SPI-Archive" button. Your guess sounds plausible, I'll give that a try - thanks for the idea!

**Comment by roysmith** [comment]
There's another example at https://en.wikipedia.org/w/index.php?title=Wikipedia:Sockpuppet_investigations/Salma_Abdallah_of_Somalia/Archive&action=history.  I remember when I archived this, there was an entry in the log about an edit conflict, but it went by too fast and I couldn't capture the details.  In any case, it ended up with 2 copies of 22 January 2021, 3 seconds apart.

Drmies later came along and left me a note on one of the copies, but that was quite a bit later.

**Comment by roysmith** [comment]
Here's another clue.  Look at Wikipedia:Sockpuppet investigations/Jjanhone/Archive.

I had just responded to your comment about blocking the /64's and did went to archive it.  I got:

```
Got Wikipedia:Sockpuppet investigations/Jjanhone
Page Wikipedia:Sockpuppet investigations/Jjanhone/Archive does not exist
Saved Wikipedia:Sockpuppet investigations/Jjanhone/Archive
Edit failed on Wikipedia:Sockpuppet investigations/Jjanhone: editconflict
Got Wikipedia:Sockpuppet investigations/Jjanhone
Got Wikipedia:Sockpuppet investigations/Jjanhone/Archive
Saved Wikipedia:Sockpuppet investigations/Jjanhone/Archive
Saved Wikipedia:Sockpuppet investigations/Jjanhone
Purged Wikipedia:Sockpuppet investigations/Jjanhone
Done!
```

and it ended up double-archiving.  So, yeah, this is smelling more and more like some kind of race condition on recently edited pages.

**Comment by GeneralNotability** [comment]
@roysmith I've implemented your suggestion in the just-released v2.5.1 - the archive loop now waits until the archive's revision ID changes before starting the next archive. Feel free to reopen if you see this again. Thanks for the help with this one!

**Comment by roysmith** [comment]
Is it the archive that you need to loop on changing, or the active page?

**Comment by GeneralNotability** [comment]
uhhhhh... *moves to reopened*

**Comment by roysmith** [comment]
Here's a slight variation on the same issue.  I went to archive Wikipedia:Sockpuppet investigations/Depeterz.  Got:

	•	Got Wikipedia:Sockpuppet investigations/Depeterz
	•	Page Wikipedia:Sockpuppet investigations/Depeterz/Archive does not exist
	•	Saved Wikipedia:Sockpuppet investigations/Depeterz/Archive
	•	Edit failed on Wikipedia:Sockpuppet investigations/Depeterz: editconflict

and 10 April 2021 got archived, but is also still in the current live case page.
